DESIGN.md

```diff
@@ -1,30 +1,273 @@
+├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┤
+ Writes resume as normal, continuing on after 5.
+│                                                                                     │
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+```
+
+### Write Record failure due to bad flash block
+
+```text
+┌ nor flash ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+ ┌─────────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────────┐
+││             │--23--------34--------45-------56--------6             │             ││
+ │             │xxxxxxxxxxxxxSabcdefghESabcdefgESabcdefghE      ~      │             │
+││             │             │             │             │             │             ││
+ └─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┘
+├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┤
+ We start with all valid data. Unknown to us, the next sector contains a bad byte that
+│ will fail to write.                                                                 │
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+
+┌ nor flash ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+ ┌─────────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────────┐
+││             │--23--------34--------45-------56--------67--------7   │             ││
+ │             │xxxxxxxxxxxxxSabcdefghESabcdefgESabcdefghESabcde~ghE   │             │
+││             │             │             │             │             │             ││
+ └─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┘
+├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┤
+ We write and perform a readback to validate. The CRC does not match due to the bad
+│ byte in the region. We attempt to start over and do a write again.                  │
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+
+┌ nor flash ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+ ┌─────────────┬─────────────┬─────────────┬─────────────┬─────────────┬─────────────┐
+││             │--23--------34--------45-------56--------67--------78--------8       ││
+ │             │xxxxxxxxxxxxxxxxxxxxxxxSabcdefgESabcdefghESabcde~ghESabcdefghE       │
+││             │             │             │             │             │             ││
+ └─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┴─4k sector───┘
+├ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┤
+ This write succeeds. We now have three good writes: 5, 6, and 8. All items earlier
+│ than 5 are popped. 7 is left as-is. An erase is scheduled as usual.                 │
```

Comment:
Again, this isn't EXACTLY how process_garbage works now, we WILL pop the bad 7 nodes now.

Reply:
@diondokter just checking: if "popping" a node fails because the bits are "stuck high" in the CRC of a pushed element, and "popping" can't write 0, is this going to cause us problems? We should maybe look at any cases where s-s ops will fail, that could cause us to early-return when we shouldn't.

(I'm not sure what flash failure modes look like typically, whether bits are more commonly "stuck high" or "stuck low", and if either of those can cause us problems).